src/beautifiers/markdown-blocks/index.coffee

```diff
@@ -0,0 +1,52 @@
+module.exports = (text, logger) ->
+  return new @Promise((resolve, reject) ->
+    extractCodeBlocks = require 'gfm-code-blocks'
+    beautifier = require '../../beautifier'
+    cleanMarkdown = text
+    codeBlocks = extractCodeBlocks(cleanMarkdown)
+    beautifyBlockPromises = []
+    for codeBlock in codeBlocks
+      codeLanguage = codeBlock.lang.match(/^[\w\d]+/)?[0]
+      logger.verbose "Beautify code block #{codeLanguage}"
+      if !codeLanguage
+        beautifyBlockPromises.push(null)
+      else
+        beautifyBlockPromises.push(new @Promise((resolve, reject) ->
+          linesInBlock = codeBlock.block.split('\n')
+          codePrefix = linesInBlock[0]
+          codeSuffix = linesInBlock[linesInBlock.length - 1]
+          code = linesInBlock.slice(1, linesInBlock.length - 1).join('\n')
+          filePath = 'fakepath.' + codeLanguage
+          grammar = atom.grammars.selectGrammar(filePath, code)
+          grammarName = grammar.name
+          allOptions = beautifier.getOptionsForPath(filePath)
+          originalStart = codeBlock.start
+          originalEnd = codeBlock.end
+          if grammarName == 'Null Grammar'
+            resolve(null)
+          else
+            beautifier.beautify(code, allOptions, grammarName, filePath)
+              .then((cleanCode) ->
+                if cleanCode
+                  resolve({
+                    block: [codePrefix, cleanCode, codeSuffix].join('\n')
+                    originalStart: originalStart
+                    originalEnd: originalEnd
+                  })
+                else
+                  resolve(null)
+              )
+              .catch((e) ->
+                logger.verbose "error while beautifying #{code.substring(0, 200)}...", e
+                resolve(null)
+              )
+          ))
+    Promise.all(beautifyBlockPromises).then((cleanCodeBlocks) ->
+      cleanCodeBlocks.reverse((b) -> !!b).forEach((codeBlock) ->
+        if (codeBlock)
+          cleanMarkdown = cleanMarkdown.substring(0, codeBlock.originalStart) + codeBlock.block + cleanMarkdown.substring(codeBlock.originalEnd)
+      )
+    ).then(->
+      resolve(cleanMarkdown)
```

Comment:
I do not like having side effects within Promises (i.e. `cleanMarkdown` is global variable and changed, instead of resolving the new value in the promise to be used later with `.then(newValue)`). Instead, please consider returning within the Promise itself. `Array.reduce` would be good here ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce ). For example:

```coffeescript
# ...
Promise.all(beautifyBlockPromises).then((cleanCodeBlocks) ->
  cleanCodeBlocks.reverse((b) -> !!b).reduce((cleanMarkdown, codeBlock) -> (
         if (codeBlock)
           return cleanMarkdown.substring(0, codeBlock.originalStart) + codeBlock.block  cleanMarkdown.substring(codeBlock.originalEnd)
        else
           return cleanMarkdown
       ), cleanMarkdown)
).then(resolve)
```

Where `).then(resolve)` is shorthand for `).then((cleanMarkdown) -> resolve(cleanMarkdown))`.

Reply:
I addressed this in a new commit, thanks for the hint The code looks cleaner indeed!